ENH: Add Arrow CSV Reader

lithomas1 opened the pull request:
- [ ] closes #23697
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

Picking up from #38370. 
Perf is OK. (StringIO/Text buffer reading gets like 20% overhead b/c of BytesIOWrapper, we should probably look into more efficient ways of doing it/cythonize it. Bytes IO performance looks good though).

Benchmarks 100k rows, 5 col
```
[  0.00%] ··· io.csv.ReadCSVEngine.time_read_bytescsv                                                                                                                             ok
[  0.00%] ··· ========= ============
                engine              
              --------- ------------
                  c       22.9±2ms  
                python    259±20ms  
               pyarrow   5.79±0.3ms 
              ========= ============

[  0.00%] ··· io.csv.ReadCSVEngine.time_read_stringcsv                                                                                                                            ok
[  0.00%] ··· ========= ============
                engine              
              --------- ------------
                  c       20.6±1ms  
                python    257±1ms   
               pyarrow   7.06±0.3ms 
              ========= ============
```

Tests will still fail tho, working on that rn.

My machine is a 2019 MBP(Intel) with 6 cores. YMMV based on the number of CPUs you have. Surprised that it doesn't quite scale linearly, maybe the file is too small to feel the benefit of all 6 cores.

lithomas1 committed:
ENH: Add pyarrow csv reader

lithomas1 committed:
merge master

lithomas1 committed:
address review comments

lithomas1 committed:
fix some tests

lithomas1 committed:
xfail/skip more

twoertwein commented:
> Picking up from #38370.
> Perf is OK. (StringIO gets like 20% overhead b/c of BytesIOWrapper, we should probably look into more efficient ways of doing it/cythonize it).

Isn't the `BytesIOWrapper` only used when the user provides a text-handle? If the user provides something path-like, `get_handle` should (hopefully) directly open it in binary mode, or?

lithomas1 commented:
Yup, that's why its only in the StringIO benchmark. Sorry for being a little ambiguous there.(``bytescsv`` is unaffected as you can see above).

lithomas1 committed:
more test xfails/skips

lithomas1 commented:
I think I've addressed all the comments on the previous PR. LMK otherwise. This is ready for a preliminary review now, will ping everyone involved once I get CI to green.(22 tests to go left, but some tests are getting stuck and timing out).

lithomas1 committed:
maybe green?

lithomas1 committed:
green?

lithomas1 commented:
as green as it gets.

lithomas1 commented:
also cc @arw2019 @xhochy

jreback reviewed (changes requested):
looks really good, thanks @lithomas1 for reviving this.

some doc-comments / requests
and 1 for code.

are all of the currently raised exceptions tested?

gfyoung commented on doc/source/user_guide/io.rst:
```diff
@@ -1622,11 +1627,19 @@ Specifying ``iterator=True`` will also return the ``TextFileReader`` object:
-back to Python if C-unsupported options are specified. Currently, C-unsupported
-options include:
+Pandas currently supports three engines, the C engine, the python engine, and an experimental
+pyarrow engine(which requires the ``pyarrow`` package). In general, the pyarrow engine is fastest
```
```suggestion
pyarrow engine (requires the ``pyarrow`` package). In general, the pyarrow engine is fastest
```

gfyoung commented on doc/source/user_guide/io.rst:
```diff
@@ -1622,11 +1627,19 @@ Specifying ``iterator=True`` will also return the ``TextFileReader`` object:
-options include:
+Pandas currently supports three engines, the C engine, the python engine, and an experimental
+pyarrow engine(which requires the ``pyarrow`` package). In general, the pyarrow engine is fastest
+on larger workloads, and is equivalent in speed to the C engine on most other workloads.
```
```suggestion
on larger workloads and is equivalent in speed to the C engine on most other workloads.
```

gfyoung commented on doc/source/user_guide/io.rst:
```diff
@@ -1622,11 +1627,19 @@ Specifying ``iterator=True`` will also return the ``TextFileReader`` object:
+on larger workloads, and is equivalent in speed to the C engine on most other workloads.
+The python engine tends to be slower than the pyarrow and C engines on most workloads. However,
+the pyarrow engine is much less robust than the C engine, which lacks a few features compared to the
+Python engines
```
```suggestion
Python engines.
```

gfyoung commented on doc/source/user_guide/io.rst:
```diff
@@ -1622,11 +1627,19 @@ Specifying ``iterator=True`` will also return the ``TextFileReader`` object:
+the pyarrow engine is much less robust than the C engine, which lacks a few features compared to the
+Python engines
+
+Where possible pandas uses the C parser (specified as ``engine='c'``), but it may fall
```
```suggestion
Where possible, pandas uses the C parser (specified as ``engine='c'``), but it may fall
```

lithomas1 committed:
Apply suggestions from code review

Co-authored-by: gfyoung <gfyoung17+GitHub@gmail.com>

lithomas1 committed:
Merge branch 'master' into enh-arrow-csv

lithomas1 commented on pandas/io/parsers/readers.py:
```diff
@@ -1043,22 +1111,25 @@ def _failover_to_python(self):
+            df = self._engine.read()
         else:
-            new_rows = len(index)
+            nrows = validate_integer("nrows", nrows)
```
#43135.

lithomas1 committed:
code review

lithomas1 commented:
Raised exceptions tested here.
https://github.com/pandas-dev/pandas/pull/43072/files#diff-2a0f93df26fa0ee1d10f216142db735ff227fcfb02f161ad165fdfcb3f767cd1R125-R146.

jreback commented:
@pandas-dev/pandas-core if any comments. Ideally we could do these in a followup.

Dr-Irv commented on doc/source/user_guide/io.rst:
```diff
@@ -1635,6 +1647,32 @@ options include:
+* ``low_memory``
+
+Specifying these options will raise a ``ValueError``.
+
```
Might want to say "Specifying these options with ``engine='pyarrow'`` will raise a ``ValueError``.

Dr-Irv commented on pandas/io/parsers/readers.py:
```diff
@@ -878,7 +939,7 @@ def _clean_options(self, options, engine):
+            if engine in ("c", "pyarrow"):
                 fallback_reason = (
                     "the 'c' engine does not support "
                     "sep=None with delim_whitespace=False"
```
Might want to update the error message here to say `f"the '{engine}' does not support "` so that it refers to the specified engine.

lithomas1 replied:
oops fixed now.

Dr-Irv reviewed (commented):
Some minor comments

lithomas1 committed:
fix typos

lithomas1 committed:
Merge branch 'pandas-dev:master' into enh-arrow-csv

jreback commented:
thanks @lithomas1 very nice!


pls create a follow up issue with checkboxes

jreback commented:
thanks @arw2019 for a lot of the initial work here!

lithomas1 commented:
#38889 
#38872

FWIW, is it possible to create an Arrow label for pyarrow things? Arrow issues are still filed under IO Parquet which is wrong. I think this might be relevant for things like the ArrowStringArray too.

jreback commented:
i created a label just now

jbrockmendel commented:
im seeing pyarrow io tests stalling out locally (and not KeyboardInterrupt-able), can anyone else on mac confirm?

lithomas1 commented:
Maybe try setting the env var ``OMP_NUM_THREADS`` to 1? We should be running these tests single threaded right now.
I'll try to reproduce when I get back on my mac, but its interesting how this is not failing on CI.

lithomas1 commented:
Ok, looking into this right now. I think I can reproduce locally, but interested in how this doesn't reproduce on CI.

jbrockmendel commented:
I fixed it locally by upgrading pyarrow
